output/nsq/outputnsq.go

```diff
@@ -0,0 +1,81 @@
+package nsq
+
+import (
+	"context"
+	"errors"
+	"github.com/nsqio/go-nsq"
+	"github.com/tsaikd/KDGoLib/version"
+	"github.com/tsaikd/gogstash/config"
+	"github.com/tsaikd/gogstash/config/goglog"
+	"github.com/tsaikd/gogstash/config/logevent"
+)
+
+// ModuleName is the name used in config file
+const ModuleName = "nsq"
+
+// OutputConfig holds the configuration json fields and internal objects
+type OutputConfig struct {
+	config.OutputConfig
+	NSQ      string `json:"nsq" yaml:"nsq"`                   // NSQd to connect to
+	Topic    string `json:"topic" yaml:"topic"`               // topic to publish to
+	InFlight uint   `json:"max_inflight" yaml:"max_inflight"` // max number of messages inflight
+
+	producer *nsq.Producer
+	ctx      context.Context
+}
+
+// DefaultOutputConfig returns an OutputConfig struct with default values
+func DefaultOutputConfig() OutputConfig {
+	return OutputConfig{
+		OutputConfig: config.OutputConfig{
+			CommonConfig: config.CommonConfig{
+				Type: ModuleName,
+			},
+		},
+		InFlight: 150,
+	}
+}
+
+// InitHandler initialize the output plugin
+func InitHandler(ctx context.Context, raw *config.ConfigRaw) (config.TypeOutputConfig, error) {
+	conf := DefaultOutputConfig()
+	err := config.ReflectConfig(raw, &conf)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(conf.NSQ) == 0 {
+		return nil, errors.New("Missing NSQ server")
+	}
+	if len(conf.Topic) == 0 {
+		return nil, errors.New("Missing topic")
+	}
+
+	conf.ctx = ctx
+	cfg := nsq.NewConfig()
+	cfg.MaxInFlight = int(conf.InFlight)
+	cfg.UserAgent = "gogstash/" + version.VERSION
+	conf.producer, err = nsq.NewProducer(conf.NSQ, cfg)
+	if err != nil {
+		return nil, err
+	}
+	go conf.nsqbackgroundtask()
```

Comment:
I'm curious, do you check if the background task doing all the things as you wish?

Reply:
During testing  I checked this by looking after the output on line 69 & 71 when I tested. I expect them to come up when I stop the program. The goal is to make sure go-nsq pushes as much as possible to the queue before the program stops.

I have not checked that go-nsq does this properly - I have only used their API as they do in their own code. t.producer.Stop() is blocking I am not supposed to get control back before the producer has stopped.

Below you see the last lines from the console when I hit CTRL+C. You can see that the the background task works as expected.

```text
2021/10/06 17:10:40 util.go:116 [info] interrupt
2021/10/06 17:10:40 INF    1 [temp/temp] stopping...
2021/10/06 17:10:40 outputnsq.go:69 [debug] outputnsq: stopping
2021/10/06 17:10:40 INF    2 stopping
2021/10/06 17:10:40 INF    2 exiting router
2021/10/06 17:10:40 INF    2 (localhost:4150) beginning close
2021/10/06 17:10:40 INF    2 (localhost:4150) readLoop exiting
2021/10/06 17:10:40 outputnsq.go:71 [debug] outputnsq: stopped
2021/10/06 17:10:40 INF    2 (localhost:4150) breaking out of writeLoop
2021/10/06 17:10:40 INF    2 (localhost:4150) writeLoop exiting
2021/10/06 17:10:40 INF    1 [temp/temp] (localhost:4150) received CLOSE_WAIT from nsqd
2021/10/06 17:10:40 INF    1 [temp/temp] (localhost:4150) beginning close
2021/10/06 17:10:40 INF    1 [temp/temp] (localhost:4150) readLoop exiting
2021/10/06 17:10:40 INF    1 [temp/temp] (localhost:4150) breaking out of writeLoop
2021/10/06 17:10:40 INF    1 [temp/temp] (localhost:4150) writeLoop exiting
2021/10/06 17:10:40 INF    2 (localhost:4150) finished draining, cleanup exiting
2021/10/06 17:10:40 INF    2 (localhost:4150) clean close complete
2021/10/06 17:10:40 INF    1 [temp/temp] (localhost:4150) finished draining, cleanup exiting
2021/10/06 17:10:40 INF    1 [temp/temp] (localhost:4150) clean close complete
2021/10/06 17:10:40 WRN    1 [temp/temp] there are 0 connections left alive
2021/10/06 17:10:40 INF    1 [temp/temp] stopping handlers
2021/10/06 17:10:40 INF    1 [temp/temp] rdyLoop exiting
```